components/places/src/api/placesapi.rs

```diff
@@ -0,0 +1,256 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+use crate::db::db::PlacesDb;
+use crate::error::*;
+use crate::history_sync::store::HistoryStore;
+use lazy_static::lazy_static;
+use rusqlite::OpenFlags;
+use std::cell::Cell;
+use std::collections::HashMap;
+use std::mem;
+use std::path::PathBuf;
+use std::sync::Arc;
+use std::sync::Mutex;
+use sync15::{telemetry, ClientInfo};
+
+#[repr(u8)]
+#[derive(Debug, Copy, Clone, PartialEq)]
+pub enum ConnectionType {
+    ReadOnly = 1,
+    ReadWrite = 2,
+    Sync = 3,
+}
+
+impl ConnectionType {
+    pub fn from_primitive(p: u8) -> Option<Self> {
+        match p {
+            1 => Some(ConnectionType::ReadOnly),
+            2 => Some(ConnectionType::ReadWrite),
+            3 => Some(ConnectionType::Sync),
+            _ => None,
+        }
+    }
+}
+
+impl ConnectionType {
+    pub fn rusqlite_flags(&self) -> OpenFlags {
+        let common_flags = OpenFlags::SQLITE_OPEN_NO_MUTEX | OpenFlags::SQLITE_OPEN_URI;
+        match self {
+            ConnectionType::ReadOnly => common_flags | OpenFlags::SQLITE_OPEN_READ_ONLY,
+            ConnectionType::ReadWrite => {
+                common_flags | OpenFlags::SQLITE_OPEN_CREATE | OpenFlags::SQLITE_OPEN_READ_WRITE
+            }
+            ConnectionType::Sync => common_flags | OpenFlags::SQLITE_OPEN_READ_WRITE,
+        }
+    }
+}
+
+// We only allow a single PlacesApi per filename.
+// XXX - probably need encryption key here too so we can do something sane
+// if we attempt to open the same file with different keys.
```

Comment:
I think there are security reasons not to let strings holding the encryption key live longer than necessary.

I also don't know that I think that supporting using an encrypted database is worth the trouble or (surprising) performance impact. It also ties us to SQLcipher in a way that's somewhat undesirable (since there are nontrivial amounts of work required to get it into desktop).